aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -419,10 +410,278 @@ impl AptosVM {
+        // If txn args are not valid, we'd still consider the multisig transaction as executed but
+        // failed. This is primarily because it's unrecoverable at this point.
+        let args = verifier::transaction_arg_validation::validate_combine_signer_and_txn_args(
+            session,
+            vec![multisig_address],
+            payload.args().to_vec(),
+            &function,
+        )?;
+        session
+            .execute_entry_function(
+                payload.module(),
+                payload.function(),
+                payload.ty_args().to_vec(),
+                args,
+                gas_meter,
+            )
+            .map_err(|e| e.into_vm_status())?;
+
+        // Resolve any pending module publishes in case the multisig transaction is deploying
+        // modules.
+        self.resolve_pending_code_publish(session, gas_meter, new_published_modules_loaded)?;
+        Ok(())
+    }
+
+    fn success_multisig_payload_cleanup<S: MoveResolverExt + StateView, SS: MoveResolverExt>(
+        &self,
+        storage: &S,
+        session: SessionExt<SS>,
+        gas_meter: &mut AptosGasMeter,
+        txn_data: &TransactionMetadata,
+        cleanup_args: Vec<Vec<u8>>,
+    ) -> Result<ChangeSetExt, VMStatus> {
+        // Charge gas for writeset before we do cleanup. This ensures we don't charge gas for
+        // cleanup writeset changes, which is consistent with outer-level success cleanup
+        // flow. We also wouldn't need to worry that we run out of gas when doing cleanup.
+        let inner_function_change_set_ext = session
+            .finish(&mut (), gas_meter.change_set_configs())
+            .map_err(|e| e.into_vm_status())?;
+        gas_meter.charge_write_set_gas_for_io(inner_function_change_set_ext.write_set().iter())?;
+        gas_meter.charge_storage_fee(
+            inner_function_change_set_ext.write_set().iter(),
+            inner_function_change_set_ext.change_set().events(),
+            txn_data.transaction_size,
+            txn_data.gas_unit_price,
+        )?;
+
+        let storage_with_changes =
+            DeltaStateView::new(storage, inner_function_change_set_ext.write_set());
+        let delta_write_set_mut = inner_function_change_set_ext
+            .delta_change_set()
+            .clone()
+            .try_into_write_set_mut(storage)
+            .expect("something terrible happened when applying aggregator deltas");
+        let delta_write_set = delta_write_set_mut
+            .freeze()
+            .map_err(|_err| VMStatus::Error(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
+        let storage_with_changes =
+            DeltaStateView::new(&storage_with_changes, &delta_write_set).into_move_resolver();
+        let resolver = self.0.new_move_resolver(&storage_with_changes);
+        let mut cleanup_session = self.0.new_session(&resolver, SessionId::txn_meta(txn_data));
+        cleanup_session.execute_function_bypass_visibility(
```

Comment:
can't we just dispatch here based upon the type of txn? it seems like the rest of this code largely is copy pasta/

Reply:
What do you mean by dispatch here based upon the type of txn? This is calling a specific private function in the multisig_account module that's meant to resolve the successfully executed multisig tx.